### ISSUE TITLE
local development, supabase has different values on client and server. 

### ISSUE THREAD

**Issue by shriharip** [supabase/auth-helpers#476]
# Bug report

- [X] I confirm this is a bug with Supabase, not with my own application. 
- [X] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I am using the nextjs "app" dir. (server components)
I am following the docs in creating the server, client and middleware clients. And the usual drill of using a provider and listener.

The relevant codes

```
utils/server.ts
import { headers, cookies } from 'next/headers'
import { createServerComponentSupabaseClient } from '@supabase/auth-helpers-nextjs'

 import type { Database } from '@/types/supabase'

 export const createClient = () =>
   createServerComponentSupabaseClient<Database>({
     headers,
     cookies,
   })

```

```
//utils/client.ts
import { createBrowserSupabaseClient } from '@supabase/auth-helpers-nextjs'

import type { Database } from '@/types/supabase'


export const createClient = () =>
  createBrowserSupabaseClient<Database>({
    supabaseUrl: process.env.LOCAL_SUPABASE_URL,
    supabaseKey: process.env.LOCAL_ANON_KEY,
  })

```
```
'use client'

// SUPABASE PROVIDER

import { createContext, useContext, useEffect, useState } from 'react'
import { createClient } from '@/app/lib/supabaseClient'

import type { Session, SupabaseClient } from '@supabase/auth-helpers-nextjs'
import { useRouter } from 'next/navigation'
import type { Database } from '@/types/supabase'

type SupabaseContext = {
	supabase: SupabaseClient<Database>
	session: Session | null
}

const Context = createContext<SupabaseContext | undefined>(undefined)

export default function SupabaseProvider({ children, sess }: { children: React.ReactNode, sess: Session }) {
	const [supabase] = useState(() => createClient())
	const router = useRouter()

	useEffect(() => {
		const { data: { subscription } } = supabase.auth.onAuthStateChange((event, session) => {
			if (session?.access_token !== sess?.access_token) {
				router.refresh()
			}
		})
		return () => subscription.unsubscribe()
	}, [])

	return (
		<Context.Provider value={{ supabase: supabase, session: sess }}>
			<>{children}</>
		</Context.Provider>
	)
}

export const useSupabase = () => {
	let context = useContext(Context);
	if (context === undefined) {
		throw new Error("useSupabase must be used inside SupabaseProvider");
	} else {
		return context;
	}
}
```

```
'use client'

//SUPABASE LISTENER

import { useRouter } from 'next/navigation'
import { useEffect } from 'react'
import { useSupabase } from './SupabaseProvider'

export default function SupabaseListener({ serverAccessToken }: { serverAccessToken: any }) {
	const { supabase } = useSupabase()
	const router = useRouter()

	useEffect(() => {
		const {
			data: { subscription },
		} = supabase.auth.onAuthStateChange((event, session) => {
			if (session?.access_token !== serverAccessToken) {
				router.refresh()
			}
		})

		return () => {
			subscription.unsubscribe()
		}
	}, [serverAccessToken, router, supabase])

	return null
}

```
## Expected behavior

Inside my auth component which uses  ```use client``` directive . I use ```useSupabase```

```
export default AuthComp..... 

	const { supabase } = useSupabase()

	console.log(supabase, 'supabase from auth')
```

The value I have on the console before the render is 
```
SupabaseClient {
  supabaseUrl: 'http://localhost:54321',
  supabaseKey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0',
  realtimeUrl: 'ws://localhost:54321/realtime/v1',
  authUrl: 'http://localhost:54321/auth/v1',
  storageUrl: 'http://localhost:54321/storage/v1',
  functionsUrl: 'http://localhost:54321/functions/v1',
  storageKey: 'supabase-auth-token',
  headers: { 'X-Client-Info': '@supabase/auth-helpers-nextjs@0.5.4' },
  auth: SupabaseAuthClient {
    stateChangeEmitters: Map(1) { '7437ed19-6d02-45ac-b82f-4778754382e5' => [Object] },
    autoRefreshTicker: Timeout {
      _idleTimeout: 10000,
      _idlePrev: [TimersList],
      _idleNext: [TimersList],
      _idleStart: 25714,
      _onTimeout: [Function (anonymous)],
      _timerArgs: undefined,
      _repeat: 10000,
      _destroyed: false,
      [Symbol(refed)]: false,
      [Symbol(kHasPrimitive)]: false,
      [Symbol(asyncId)]: 41733,
      [Symbol(triggerId)]: 41716,
      [Symbol(kResourceStore)]: [Object],
      [Symbol(kResourceStore)]: [Object],
      [Symbol(kResourceStore)]: undefined
    },
    visibilityChangedCallback: null,
    refreshingDeferred: null,
    initializePromise: Promise {
      [Object],
      [Symbol(async_id_symbol)]: 41678,
      [Symbol(trigger_async_id_symbol)]: 41652,
      [Symbol(kResourceStore)]: [Object],
      [Symbol(kResourceStore)]: [Object],
      [Symbol(kResourceStore)]: undefined
    },
    detectSessionInUrl: true,
    broadcastChannel: null,
    inMemorySession: null,
    storageKey: 'supabase-auth-token',
    autoRefreshToken: true,
    persistSession: true,
    storage: {
      getItem: [Function: getItem],
      setItem: [Function: setItem],
      removeItem: [Function: removeItem]
    },
    admin: GoTrueAdminApi {
      url: 'http://localhost:54321/auth/v1',
      headers: [Object],
      fetch: [Function (anonymous)],
      mfa: [Object]
    },
    url: 'http://localhost:54321/auth/v1',
    headers: {
      Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0',
      apikey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0',
      'X-Client-Info': '@supabase/auth-helpers-nextjs@0.5.4'
    },
    fetch: [Function (anonymous)],
    mfa: {
      verify: [Function: bound _verify],
      enroll: [Function: bound _enroll],
      unenroll: [Function: bound _unenroll],
      challenge: [Function: bound _challenge],
      listFactors: [Function: bound _listFactors],
      challengeAndVerify: [Function: bound _challengeAndVerify],
      getAuthenticatorAssuranceLevel: [Function: bound _getAuthenticatorAssuranceLevel]
    }
  },
  fetch: [Function (anonymous)],
  realtime: RealtimeClient {
    accessToken: null,
    channels: [],
    endPoint: 'ws://localhost:54321/realtime/v1/websocket',
    headers: { 'X-Client-Info': '@supabase/auth-helpers-nextjs@0.5.4' },
    params: {
      apikey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0'
    },
    timeout: 10000,
    transport: [Function: W3CWebSocket],
    heartbeatIntervalMs: 30000,
    heartbeatTimer: undefined,
    pendingHeartbeatRef: null,
    ref: 0,
    logger: [Function: noop],
    conn: null,
    sendBuffer: [],
    serializer: Serializer { HEADER_LENGTH: 1 },
    stateChangeCallbacks: { open: [], close: [], error: [], message: [] },
    eventsPerSecondLimitMs: 100,
    inThrottle: false,
    reconnectAfterMs: [Function (anonymous)],
    encode: [Function (anonymous)],
    decode: [Function: bound decode],
    reconnectTimer: Timer {
      callback: [Function (anonymous)],
      timerCalc: [Function (anonymous)],
      timer: undefined,
      tries: 0
    }
  },
  rest: PostgrestClient {
    url: 'http://localhost:54321/rest/v1',
    headers: { 'X-Client-Info': '@supabase/auth-helpers-nextjs@0.5.4' },
    schema: 'public',
    fetch: [Function (anonymous)]
  }
} supabase from auth
```

However after the render  console.log on the browser shows me the project details in the browser console.

```
{
  "supabaseUrl": "https://xxxxxxxxkajwrtbmxsat.supabase.co",
  "supabaseKey": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM",
  "realtimeUrl": "wss://Xxxxxxxxxxxxsat.supabase.co/realtime/v1",
  "authUrl": "https://xxxxxxxxmxsat.supabase.co/auth/v1",
  "storageUrl": "https://jXxxxxxxxxx.supabase.co/storage/v1",
  "functionsUrl": "https://Xxxxxxxxxxt.functions.supabase.co",
  "storageKey": "supabase-auth-token",
  "headers": {
    "X-Client-Info": "@supabase/auth-helpers-nextjs@0.5.4"
  },
  "auth": {
    "stateChangeEmitters": {},
    "autoRefreshTicker": 2994,
    "refreshingDeferred": null,
    "initializePromise": {},
    "detectSessionInUrl": true,
    "broadcastChannel": {},
    "inMemorySession": null,
    "storageKey": "supabase-auth-token",
    "autoRefreshToken": true,
    "persistSession": true,
    "storage": {},
    "admin": {
      "url": "https://Xxxxxxxxxxxat.supabase.co/auth/v1",
      "headers": {
        "Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM",
        "apikey": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM",
        "X-Client-Info": "@supabase/auth-helpers-nextjs@0.5.4"
      },
      "mfa": {}
    },
    "url": "https://jargqamakajwrtbmxsat.supabase.co/auth/v1",
    "headers": {
      "Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM",
      "apikey": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM",
      "X-Client-Info": "@supabase/auth-helpers-nextjs@0.5.4"
    },
    "mfa": {}
  },
  "realtime": {
    "accessToken": null,
    "channels": [],
    "endPoint": "wss://jargxxxxxxxxxat.supabase.co/realtime/v1/websocket",
    "headers": {
      "X-Client-Info": "@supabase/auth-helpers-nextjs@0.5.4"
    },
    "params": {
      "apikey": "eyJhbGciOiJIUzI1NiIsInRxxxxpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImphcmdxYW1ha2Fqd3J0Ym14c2F0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzY3OTAwMzYsImV4cCI6MTk5MjM2NjAzNn0.hk7a3YmKqDPp9T6ZYFc1QUM3XS4O2iu67mKlQchLGdM"
    },
    "timeout": 10000,
    "heartbeatIntervalMs": 30000,
    "pendingHeartbeatRef": null,
    "ref": 0,
    "conn": null,
    "sendBuffer": [],
    "serializer": {
      "HEADER_LENGTH": 1
    },
    "stateChangeCallbacks": {
      "open": [],
      "close": [],
      "error": [],
      "message": []
    },
    "eventsPerSecondLimitMs": 100,
    "inThrottle": false,
    "reconnectTimer": {
      "tries": 0
    }
  },
  "rest": {
    "url": "https://jargxxxxxxxxxat.supabase.co/rest/v1",
    "headers": {
      "X-Client-Info": "@supabase/auth-helpers-nextjs@0.5.4"
    },
    "schema": "public"
  }
}
```

I would expect to have the same localhost url in the browser also. 

Because of this, I am not able to test the login and auth in the local  developments since the email verification does not hit the local server. Nor does the db gets the value and it keeps going into the production database.

## System information

- OS: [e.g. macOS, Windows]
- macOS 12.6 Monterey
- Browser (if applies) [e.g. chrome, safari]
- chrome and Firefox
- Version of supabase-js: [e.g. 6.0.2]
- supabase 2.12.0
- Version of Node.js: [e.g. 10.10.0]
node 17.3

**Comment by silentworks** [comment]
Please provide a minimal reproducible repo with the issue, its really hard to follow with only code snippets in the messages.
